### PR TITLE
Home work 3 Locators

### PR DESCRIPTION
**Pull request by YanaPokulita**

### CODE REVIEW

File: tests/yana_pokulita/locators.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+from selenium.webdriver.common.by import By
+
+
+class Locators:
+    # check cart indicator
+
+    # (By.ID, "shopping_cart_container")
+    shoppingCart = (By.CSS_SELECTOR, ".primary_header .shopping_cart_container")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:8]: python naming conventions usually are:

some_var_name
SOME_CONSTANT_NAME
SomeClassName

File: tests/yana_pokulita/locators.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    # (By.ID, "finish")
+    FinishBtn = (By.CSS_SELECTOR, ".cart_footer .cart_button[data-test='finish'")
+
+    # go back to products
+
+    # (By.ID, "back-to-products")
+    BackHomeBtn = (By.CSS_SELECTOR, '#back-to-products')
+
+
+class ForConfTest:
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:46]: locators should not care, what file uses them, they should represent web pages , and give understanding on what web page or in what element on that page they are located

File: tests/yana_pokulita/locators.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    LastName = (By.CSS_SELECTOR, ".form_group .form_input[placeholder='Last Name'")
+    # (By.ID, "postal-code")
+    PostalCode = (By.CSS_SELECTOR, ".form_group .form_input[placeholder='Zip/Postal Code'")
+    # (By.ID, "continue")
+    btnContinue = (By.CSS_SELECTOR, ".checkout_buttons .submit-button[name='continue'")
+
+    # order submission
+
+    # (By.ID, "finish")
+    FinishBtn = (By.CSS_SELECTOR, ".cart_footer .cart_button[data-test='finish'")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:38]: its simply by id

(By.CSS_SELECTOR, "#finish")

File: tests/yana_pokulita/locators.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    # fill the form
+
+    # (By.ID, "first-name")
+    FirstName = (By.CSS_SELECTOR, ".form_group .form_input[placeholder='First Name'")
+    # (By.ID, "last-name")
+    LastName = (By.CSS_SELECTOR, ".form_group .form_input[placeholder='Last Name'")
+    # (By.ID, "postal-code")
+    PostalCode = (By.CSS_SELECTOR, ".form_group .form_input[placeholder='Zip/Postal Code'")
+    # (By.ID, "continue")
+    btnContinue = (By.CSS_SELECTOR, ".checkout_buttons .submit-button[name='continue'")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:33]: locator not closed appropriately with ]

File: tests/yana_pokulita/locators.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    # delete element from the cart
+
+    # (By.CSS_SELECTOR, "button#remove-sauce-labs-backpack")
+    Remove1stItemBtn = (By.XPATH, "//*[@id='remove-sauce-labs-backpack']")
+
+    # fill the form
+
+    # (By.ID, "first-name")
+    FirstName = (By.CSS_SELECTOR, ".form_group .form_input[placeholder='First Name'")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:27]: rely on ID

(By.CSS_SELECTOR, "#first-name")

File: tests/yana_pokulita/locators.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    # (By.ID, "shopping_cart_container")
+    shoppingCart = (By.CSS_SELECTOR, ".primary_header .shopping_cart_container")
+    # (By.XPATH, ".//*[contains(@class, 'shopping_cart_badge')]")
+    shoppingCartBadge = (By.CSS_SELECTOR, ".shopping_cart_link .shopping_cart_badge[data-test='shopping-cart-badge']")
+
+    # check number of items in the cart
+
+    # (By.CSS_SELECTOR, ".cart_list .cart_item")
+    CartItems = (By.XPATH, ".//*[@class='cart_item']")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:15]: use contains to rely on class name using xpath

File: tests/yana_pokulita/locators.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
+from selenium.webdriver.common.by import By
+
+
+class Locators:
+    # check cart indicator
+
+    # (By.ID, "shopping_cart_container")
+    shoppingCart = (By.CSS_SELECTOR, ".primary_header .shopping_cart_container")
+    # (By.XPATH, ".//*[contains(@class, 'shopping_cart_badge')]")
+    shoppingCartBadge = (By.CSS_SELECTOR, ".shopping_cart_link .shopping_cart_badge[data-test='shopping-cart-badge']")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:10]: this is how it looks if to translate from xpath to css selector

(By.CSS_SELECTOR, ".shopping_cart_badge")

**Review by kudinov-fedor — CHANGES_REQUESTED** [review]
rework locators

File: tests/yana_pokulita/locators.py (hunk 2 above)
Comment (YanaPokulita) [reply tests/yana_pokulita/locators.py:46]: tbh not sure what this means

File: tests/yana_pokulita/locators.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
+from selenium.webdriver.common.by import By
+
+
+class Locators:
+    # check cart indicator
+
+    # (By.ID, "shopping_cart_container")
+    ShoppingCart = (By.CSS_SELECTOR, ".primary_header .shopping_cart_container")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:8]: need to fix

File: tests/yana_pokulita/locators.py (hunk 6 above)
Comment (fkudinov-mn) [reply tests/yana_pokulita/locators.py:15]: need to fix

File: tests/yana_pokulita/locators.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    # (By.XPATH, ".//*[contains(@class, 'shopping_cart_badge')]")
+    ShoppingCartBadge = (By.CSS_SELECTOR, ".shopping_cart_badge")
+
+    # check number of items in the cart
+
+    # (By.CSS_SELECTOR, ".cart_list .cart_item")
+    CartItems = (By.XPATH, ".//*[@class='cart_item']")
+    # (By.ID, "checkout")
+    CheckOutBtn = (By.CSS_SELECTOR, ".cart_footer .checkout_button[name='checkout']")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:17]: need to fix

File: tests/yana_pokulita/locators.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+    # delete element from the cart
+
+    # (By.CSS_SELECTOR, "button#remove-sauce-labs-backpack")
+    Remove1stItemBtn = (By.XPATH, "//*[@id='remove-sauce-labs-backpack']")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:22]: add tag name to xpath

File: tests/yana_pokulita/locators.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    LastName = (By.CSS_SELECTOR, "#last-name")
+    # (By.ID, "postal-code")
+    PostalCode = (By.CSS_SELECTOR, "#postal-code")
+    # (By.ID, "continue")
+    btnContinue = (By.CSS_SELECTOR, "#continue")
+
+    # order submission
+
+    # (By.ID, "finish")
+    FinishBtn = (By.CSS_SELECTOR, "#finish")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:38]: add also xpath as a comment

File: tests/yana_pokulita/locators.py (hunk 2 above)
Comment (fkudinov-mn) [reply tests/yana_pokulita/locators.py:46]: you should have locators grouped as per page and as per page elements, but not as per project files. e.g:


```
class LoginPage:
  TXT_NAME = ...
  TXT_PASSWORD = ...
  BTN_SUBMIT = ...


class CatalogPage:
    
    CART_ICON = ...
    
    class Item:
         SELF = ...        # locator for all items in table
         BTN_TO_CART = ...     # relative locator from SELF
         TXT_PRICE = ....




# usage

items = session.find_elements(*CatalogPage.Item.SELF)
items[3].find_element(*CatalogPage.Item.BTN_TO_CART).click()
```




you do not need to see the page, locators suggest you how page looks like, and look like elements tree

File: tests/yana_pokulita/locators.py [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+    LoginBtn = (By.NAME, "login-button")
+
+    # adding elements
+
+    # (By.CSS_SELECTOR, ".inventory_list .inventory_item")
+    Elements = (By.XPATH, "//*[@class='inventory_item_description']//div[2]")
+    # (By.XPATH, ".//*[@class='pricebar']//button")
+    Element1 = (By.CSS_SELECTOR, ".pricebar .btn_inventory")
+    # (By.XPATH, ".//*[@class='pricebar']//button")
+    Element3 = (By.CSS_SELECTOR, ".pricebar .btn_inventory")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:63]: why you have identical locators here

File: tests/yana_pokulita/locators.py [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    Password = (By.NAME, "password")
+    # (By.ID, "login-button")
+    LoginBtn = (By.NAME, "login-button")
+
+    # adding elements
+
+    # (By.CSS_SELECTOR, ".inventory_list .inventory_item")
+    Elements = (By.XPATH, "//*[@class='inventory_item_description']//div[2]")
+    # (By.XPATH, ".//*[@class='pricebar']//button")
+    Element1 = (By.CSS_SELECTOR, ".pricebar .btn_inventory")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:61]: provide identical xpath locator, 
locators do not match

File: tests/yana_pokulita/locators.py [hunk 14]
```diff
@@ -1,11 +1,11 @@
 from selenium.webdriver.common.by import By
 
 
-class Locators:
+class Cart:
     # check cart indicator
 
     # (By.ID, "shopping_cart_container")
-    ShoppingCart = (By.CSS_SELECTOR, ".primary_header .shopping_cart_container")
+    ShoppingCart = (By.CSS_SELECTOR, ".shopping_cart_container")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:8]: ShoppingCart = (By.CSS_SELECTOR, "#shopping_cart_container")

File: tests/yana_pokulita/locators.py [hunk 15]
```diff
@@ -14,12 +14,12 @@ class Locators:
     # (By.CSS_SELECTOR, ".cart_list .cart_item")
     CartItems = (By.XPATH, ".//*[@class='cart_item']")
     # (By.ID, "checkout")
-    CheckOutBtn = (By.CSS_SELECTOR, ".cart_footer .checkout_button[name='checkout']")
+    CheckOutBtn = (By.CSS_SELECTOR, ".checkout_button[name='checkout']")
```
Comment (fkudinov-mn) [inline tests/yana_pokulita/locators.py:17]: (By.CSS_SELECTOR, "#checkout")

Comment (YanaPokulita) [reply tests/yana_pokulita/locators.py]: changed to suggested

File: tests/yana_pokulita/locators.py (hunk 14 above)
Comment (YanaPokulita) [reply tests/yana_pokulita/locators.py]: changed to suggested

File: tests/yana_pokulita/locators.py (hunk 6 above)
Comment (kudinov-fedor) [reply tests/yana_pokulita/locators.py:15]: CartItems = (By.XPATH, ".//*[contains(@class, 'cart_item')]")

**Review by kudinov-fedor — APPROVED** [review]

File: tests/yana_pokulita/locators.py [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    LoginBtn = (By.NAME, "login-button")
+
+
+class ProductPage:
+    # adding elements
+
+    # (By.CSS_SELECTOR, ".inventory_list .inventory_item")
+    Elements = (By.XPATH, "//div[contains(@class, 'inventory_list')]//div[contains(@class, 'inventory_item_description')]")
+    # (By.XPATH, ".//*[@class='pricebar']//button")
+    Element = (By.CSS_SELECTOR, ".pricebar button")
```
Comment (kudinov-fedor) [inline tests/yana_pokulita/locators.py:64]: from name completely not clear that it is add to cart button

File: tests/yana_pokulita/locators.py [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    Password = (By.NAME, "password")
+    # (By.ID, "login-button")
+    LoginBtn = (By.NAME, "login-button")
+
+
+class ProductPage:
+    # adding elements
+
+    # (By.CSS_SELECTOR, ".inventory_list .inventory_item")
+    Elements = (By.XPATH, "//div[contains(@class, 'inventory_list')]//div[contains(@class, 'inventory_item_description')]")
```
Comment (kudinov-fedor) [inline tests/yana_pokulita/locators.py:62]: its not elements - its element, all depends on how you use it

.find_element(*Elements)
.find_elements(*Elements)

**Review by kudinov-fedor — APPROVED** [review]

**Review by kudinov-fedor — APPROVED** [review]

**Pull request merged**